utilities/imgcat

```diff
@@ -50,37 +52,68 @@ function b64_decode() {
 #   base64contents: Base64-encoded contents
-#   print_filename: If non-empty, print the filename
-#                   before outputting the image
+#   print_filename: 0 or 1, if set to 1, print the filename after outputting the image
+#   width: set output width of the image in character cells, pixels or percent
+#   height: set output height of the image in character cells, pixels or percent
+#   preserve_aspect_ratio: 0 or 1, if set to 1, fill the specified width and height as much as possible without stretching the image
 function print_image() {
     print_osc
-    printf '1337;File='
-    if [[ -n $1 ]]; then
-        printf "name=%s;" "$(printf "%s" "$1" | b64_encode)"
-    fi
-
-    printf "%s" "$3" | b64_decode | wc -c | awk '{printf "size=%d",$1}'
-    printf ";inline=%s" "$2"
-    printf ":"
-    printf "%s" "$3"
+    printf "1337;File=inline=%s" "$2"
+    printf ";size=%d" $(printf "%s" "$3" | b64_decode | wc -c)
+    [ -n "$1" ] && printf ";name=%s" "$(printf "%s" "$1" | b64_encode)"
+    [ -n "$5" ] && printf ";width=%s" "$5"
+    [ -n "$6" ] && printf ";height=%s" "$6"
+    [ -n "$7" ] && printf ";preserveAspectRatio=%s" "$7"
+    printf ":%s" "$3"
     print_st
     printf '\n'
-    if [[ -n $4 ]]; then
-        echo "$1"
-    fi
+    [ "$4" == "1" ] && echo "$1"
+    has_image_displayed=t
 }
 
 function error() {
     echo "ERROR: $*" 1>&2
 }
 
+function errcho() {
+    echo "$@" >&2
+}
+
 function show_help() {
-    echo "Usage: imgcat [-p] filename ..." 1>&2
-    echo "   or: cat filename | imgcat" 1>&2
+    errcho
+    errcho "Usage: imgcat [-p] [-q] [-W width] [-H height] [-r] [-s] [-u] [-f] filename ..."
+    errcho "       cat filename | imgcat [-W width] [-H height] [-r] [-s]"
+    errcho
+    errcho "Display images inline in the iTerm2 using Inline Images Protocol"
+    errcho
+    errcho "Options:"
+    errcho
+    errcho "    -h, --help                      Display help message"
+    errcho "    -p, --print                     Enable printing of filename or URL after displaying each image"
+    errcho "    -q, --quiet                     Be quiet, disable filename printing (default)"
+    errcho "    -u, --url                       Interpret following filename arguments as remote URLs and fetch they using curl"
+    errcho "    -f, --file                      Interpret following filename arguments as regular Files (default)"
+    errcho "    -r, --preserve-aspect-ratio     When scaling image preserve its original aspect ratio (default)"
+    errcho "    -s, --stretch                   Stretch image to specified width and height, this option is opposite to -r"
+    errcho "    -W, --width N                   Set image width to N character cells, pixels or percent (see below)"
+    errcho "    -H, --height N                  Set image height to N character cells, pixels or percent (see below)"
+    errcho
+    errcho "    The width and height are given as a number N followed by a unit:"
```

Comment:
We should also document that if you don't specify a width or height, an appropriate value will be chosen automatically, and that it's safe to specify one without the other.